python/ray/experimental/multiprocessing/pool.py

```diff
@@ -0,0 +1,488 @@
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+
+from multiprocessing import TimeoutError
+import os
+import time
+import random
+import threading
+import math
+import queue
+import copy
+
+import ray
+
+
+class PoolTaskError(Exception):
+    def __init__(self, underlying):
+        self.underlying = underlying
+
+
+class ResultThread(threading.Thread):
+    def __init__(self,
+                 object_ids,
+                 callback=None,
+                 error_callback=None,
+                 total_object_ids=None):
+        threading.Thread.__init__(self)
+        self._done = False
+        self._got_error = False
+        self._lock = threading.Lock()
+        self._object_ids = []
+        self._num_ready = 0
+        self._results = []
+        self._ready_index_queue = queue.Queue()
+        self._callback = callback
+        self._error_callback = error_callback
+        self._total_object_ids = total_object_ids if total_object_ids else len(
+            object_ids)
+        self._indices = {}
+        self._new_object_ids = queue.Queue()
```

Comment:
Is this queue intended to be thread-safe? If not, might the fact that `Queue` is thread-safe be a performance concern? I'm not sure, but if it is, something else might be better (apparently it uses `deque` internally which isn't synchronized).

Reply:
It is intended to be thread-safe - the caller might add object IDs to the queue as it's being read by this thread. Will add a comment.